src/infra/dex/balancer/dto.rs

```diff
@@ -1,56 +1,243 @@
+                .ok_or(Error::MissingDecimals(order.buy)),
+            order::Side::Sell => tokens
+                .decimals(&order.sell)
+                .ok_or(Error::MissingDecimals(order.sell)),
+        }?;
+        let variables = Variables {
+            call_data_input: CallDataInput {
+                deadline: swap_deadline,
+                receiver: contract_address.0,
+                sender: contract_address.0,
+                slippage_percentage: slippage.as_factor().clone(),
+            },
+            chain: Chain::from_domain(chain_id)?,
+            query_batch_swap,
+            swap_amount: HumanReadableAmount::from_decimal_units(
+                &order.amount.get(),
+                token_decimals,
+            ),
+            swap_type: SwapType::from_domain(order.side),
+            token_in: order.sell.0,
+            token_out: order.buy.0,
+            use_protocol_version: Some(VaultVersion::V2.into()),
+        };
+        Ok(Self {
+            query: QUERY,
+            variables,
+        })
+    }
+}
+
+/// Refers to the SOR API V3's `AmountHumanReadable` type and represents a token
+/// amount without decimals.
 #[serde_as]
-#[derive(Debug, Serialize)]
-#[serde(rename_all = "camelCase")]
-pub struct Query {
-    /// The sell token to quote.
-    pub sell_token: H160,
-    /// The buy token to quote.
-    pub buy_token: H160,
-    /// The order kind to use.
-    pub order_kind: OrderKind,
-    /// The amount to quote
-    ///
-    /// For sell orders this is the exact amount of sell token to trade, for buy
-    /// orders, this is the amount of buy tokens to buy.
-    #[serde_as(as = "serialize::U256")]
-    pub amount: U256,
-    /// The current gas price estimate used for determining how the trading
-    /// route should be split.
-    #[serde_as(as = "serialize::U256")]
-    pub gas_price: U256,
+#[derive(Debug, Default, PartialEq)]
+pub struct HumanReadableAmount {
+    value: BigDecimal,
+    decimals: BigDecimal,
 }
 
-impl Query {
-    pub fn from_domain(order: &dex::Order, gas_price: auction::GasPrice) -> Self {
+impl HumanReadableAmount {
+    /// Convert a `U256` amount to a human form.
+    pub fn from_decimal_units(units: &U256, decimals: u8) -> HumanReadableAmount {
+        let decimals: BigDecimal = BigInt::from(10).pow(decimals as u32).into();
         Self {
-            sell_token: order.sell.0,
-            buy_token: order.buy.0,
-            order_kind: match order.side {
-                order::Side::Buy => OrderKind::Buy,
-                order::Side::Sell => OrderKind::Sell,
-            },
-            amount: order.amount().amount,
-            gas_price: gas_price.0 .0,
+            value: u256_to_big_decimal(units) / &decimals,
+            decimals,
         }
     }
+
+    pub fn value(&self) -> &BigDecimal {
+        &self.value
+    }
+
+    /// Convert the human readable amount to a `U256` with 18 decimals.
+    pub fn to_decimal_units(&self) -> Option<U256> {
```

Comment:
This is could simply return the raw U256 value.
```suggestion
    pub fn as_wei(&self) -> U256 {
```